### DISCUSSION TITLE
Geospatial types support (initial version - `POINTS` only)

### DISCUSSION THREAD

**Discussion by snork-alt** [getdozer/dozer#877] [Feature Requests]
We need to support geospatial types and functions in Dozer. The work will consist of the following:

- Extend `dozer-types` to support geospatial types
- Allow creation and casting of geospatial types
- Include support in connectors for geospatial-types conversions
- Include support for geospatial types in the cache writer
- Implement various geospatial functions
- Update all relevant documentation

For the implementation, we will use the Rust Geo library (https://docs.rs/geo/latest/geo/)
The initial version will have limited support for geo data types. We will extend it later after we will introduce the ARRAY type

### Types support

We shall extend the `dozer-types` with the following types mapped 1 to 1 with the Rust Geo types:

- `Coord`
- `Point`
-  `Polygon`

## Creation

The creation of the above types should be supported through SQL functions. In particular:

- COORD(FLOAT, FLOAT)
- POINT(COORD)
- POLYGON(COORDS, ...)

## Functions

DISTANCE(POINT, POINT, [GEODESIC | HAVERSINE | VINCENTY]) - defaults to GEODESIC
CONTAINS(POLYGON, POINT | POLYGON)

**Comment by snork-alt** [comment]
@mediuminvader @karolisg

**Comment by karolisg** [comment]
@snork-alt What is the difference between `Point` and `Coord` types? Do we really need two different types, which represent the same data?

**Reply by mediuminvader** [reply]
it's based on mapping the rust geo implementation, quoting the crate docs:

> [edit: Coord is] A lightweight struct used to store coordinates on the 2-dimensional Cartesian plane.
Unlike Point (which in the future may contain additional information such as an envelope, a precision model, and spatial reference system information), a Coord only contains ordinate values and accessor methods.

**Reply by karolisg** [reply]
Got it. Now it is clear for me.

**Comment by chubei** [comment]
I don't think we need the `Coord` type. Because coord is the underlying representation of `Point`, `Polygon`, etc. It's not a geo type.

**Reply by mediuminvader** [reply]
Good Point 😆

**Reply by karolisg** [reply]
In that case, do we need `COORD` function in SQL? Can we use `POINT(FLOAT, FLOAT)`?

**Reply by mediuminvader** [reply]
Right, let’s go with POINT(FLOAT,FLOAT)

**Comment by karolisg** [comment]
@snork-alt @mediuminvader 

What should be the result of function POINT(FLOAT, FLOAT) and DISTANCE(POINT, POINT) when one or both arguments are null? I checked several implementations and it looks like the `NULL` result is it most common result.

Also, do we need to implement support for geo types from snowflake in the initial version?

**Reply by v3g42** [reply]
Would be nice to consider indexing these types on cache as well. Though it is not a high priority item right now.
